docs/contributing/999-rfcs/2024-02-15-resource-pinning.md

```diff
@@ -0,0 +1,239 @@
+
+### Cases
+
+For the following cases, assume a user has expressed their intent to protect a resource from deletion/moves somehow.
+The following cases should be handled by this RFC:
+
+- A pinned resource's id was changed in source (using `as`)
+- A pinned resource was moved to a different path in the resource tree (e.g. placed inside a different class)
+- An update to wing itself changed a pinned resource's path
+- An update to a dependency changed the id's within a resource it exports, and the user has pinned an instance of that resource
+
+## User Journey
+
+Sally Stateful creates a wing file with a bucket
+
+```wing
+// main.w
+bring cloud;
+
+new cloud.Bucket() as "CoolBucket";
+```
+
+They run `wing compile` and then `wing test`. Nice, looks like everything works.
+
+Running `wing compile -t tf-aws` also succeeds with the following output and a new file called main.w.pin.json appears:
+
+```shell
+# (output)
+[+pin] /root/CoolBucket
+```
+
+```json
+// main.w.pin.json
+{
+  "version": "1",
+  "pinned": {
+    "tf-aws": {
+      "/root/CoolBucket": {
+        "type": "aws_s3_bucket",
+        "uniqueId": "CoolBucket",
+        "path": "/root/CoolBucket",
+      }
+    }
+  }
+}
+```
+
+Sally deploys this to AWS and everything is great.
+
+Some time later, Sally decided to move the bucket into a new abstraction.
+
+```wing
+// main.w
+bring cloud;
+
+class MyBucket {
+  bucket: cloud.Bucket;
+
+  new() {
+    this.bucket = new cloud.Bucket();
+  }
+}
+
+let myBucket = new MyBucket();
+
+```
+
+`wing compile` and `wing test` still work as expected.
+
+`wing compile -t tf-aws` however fails with an error message
+
+```shell
+# (output)
+[-pin] /root/CoolBucket
+[+pin] /root/MyBucket/cloud.Bucket
+
+Error: The following pin has been moved or deleted:
+  - /root/CoolBucket
+
+If moved, update "path" property in the pinfile to the new path.
+If deleted, remove the entry from the pinfile.
+```
+
+For now, Sally wants to avoid the destructive move of the bucket so she updates the `path` in the pinfile.
+
+```json
+// main.w.pin.json
+{
+  "version": "1",
+  "pinned": {
+    "tf-aws": {
+      "/root/CoolBucket": {
+        "type": "aws_s3_bucket",
+        "uniqueId": "CoolBucket",
+        "path": "/root/MyBucket/cloud.Bucket"
+      }
+    }
+  }
+}
+```
+
+Running `wing compile -t tf-aws` now succeeds and nothing is deleted or moved in the resulting compiled output.
+
+Later, Sally decides the bucket has no need to be so protected, so they remove the default pin on `cloud.Bucket`:
+
+```wing
+// main.w
+bring cloud;
+
+class MyBucket {
+  bucket: cloud.Bucket;
+
+  new() {
+    this.bucket = new cloud.Bucket();
+    this.bucket.pinned = false;
+  }
+}
+
+let myBucket = new MyBucket();
+```
+
+Running `wing compile -t tf-aws` automatically updates the pinfile (because the resource is otherwise unchanged):
+
+```shell
+# (output)
+[-pin] /root/CoolBucket
```

Comment:
```suggestion
[unpinned] /root/CoolBucket
```

Reply:
I like the simple verb, it's shorter (IMO `unpinned` is kinda an icky looking word) and makes it clear that the CLI just did an action for you. Not a deal breaker for me though.